src/basic/path-util.c

```diff
@@ -703,47 +721,64 @@ int find_executable_full(const char *name, const char *root, bool use_path_envva
                 if (r == 0)
                         break;
 
-                if (!path_is_absolute(element))
-                        continue;
+                if (find_executable_from_path(name, root, element, &last_error, ret_filename, ret_fd, fd))
+                        return 0;
+        }
 
-                if (!path_extend(&element, name))
-                        return -ENOMEM;
+        return last_error;
+}
 
-                if (root) {
-                        char *path_name;
+bool find_executable_from_path(
```

Comment:
There's no reason for `last_error_addr` to be an argument here. If this function returns an `int` you can just return the error code if there is one or 0 on success. If you need a `last_error` variable just declare it locally in this function.